llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -183,29 +183,35 @@ void PostPipeliner::computeLoopCarriedParameters() {
   }
 
   // Propagate Earliest upstream, initialize Latest
+  // Unrestricted: last cycle of last stage
+  const int Latest = NCopies * II - 1;
   for (int K = 0; K < NInstr; K++) {
     const int K2 = K + NInstr;
     const int Earliest = Info[K2].Earliest - II;
     Info[K].Earliest = std::max(Info[K].Earliest, Earliest);
-    // Unrestricted: Beyond the last stage.
-    Info[K].Latest = NCopies * II;
+    Info[K].Latest = Latest;
+    Info[K2].Latest = Latest;
   }
-  // Propagate Latest upstream. Latest is the latest
-  // that is admissible for Earliest to be achievable within II
-  for (int K = 0; K < NInstr; K++) {
-    const int K2 = K + NInstr;
-    const int Earliest = Info[K2].Earliest;
-    const auto &SU = DAG->SUnits[K2];
-    for (auto &Dep : SU.Preds) {
-      const auto *Pred = Dep.getSUnit();
-      // Any predecessor in the first iteration
-      int K1 = Pred->NodeNum;
-      if (K1 < NInstr) {
-        const int Latest = Earliest - Dep.getSignedLatency();
-        Info[K1].Latest = std::min(Info[K1].Latest, Latest);
+
+  // Compute Latest. Use a fixpoint loop, because plain reversed
+  // order may not be topological for predecessors
+  bool Changed = true;
+  while (Changed) {
+    Changed = false;
+    for (int K = NInstr - 1; K >= 0; K--) {
+      SUnit &SU = DAG->SUnits[K];
+      const int Latest = Info[K].Latest;
+      for (auto &Dep : SU.Preds) {
+        int P = Dep.getSUnit()->NodeNum;
+        int NewLatest = Latest - Dep.getSignedLatency();
+        if (NewLatest < Info[P].Latest) {
```

Comment:
I would be happy with a comment. I think this means that the predecessor could be scheduled too late and that would then delay `SU`, which would then not meet its `Latest`.

Also: Could we end up with a predecessor that has a lower `Latest` than `Earliest`? If not, could we add an assert?

Reply:
This computation of Latest mimics that of Earliest. If a predecessor pushes SU too far, the schedule doesn't fit in the given linear schedule.

Latest can be smaller then Earliest, and this would say statically that the targeted linear schedule is too short.
Note that we move to an end-of-linear-schedule-relative representation of Latest, and we round the length up to a multiple of II to make Latest >= Earliest.

Also, we are going to use [Earliest, Latest] to fit it in the modulo scoreboard. If that range becomes empty, we fail.